src/types.md

```diff
@@ -236,3 +236,311 @@ fn main() {
+      Self::Growing(growing) => growing.advance(),
+      Self::Flowering(flowering) => flowering.advance(),
+      Self::Dead(dead) => dead.advance(),
+    }
+  }
+}
+
+// we should probably find a way to inject some sun and water into this
+// state machine or things are not looking rosy
+```
+
+## When should I use `Rc` versus `Arc`?
+
+Never expose `Rc` in APIs. `Arc` is the right choice nearly always. (You
+might want to look at the [archery crate](https://docs.rs/archery/latest/archery/)).
+
+## What should I do instead of inheritance?
+
+Use composition. Sometimes this results in more boilerplate, but it avoids
+a raft of complexity.
+
+Specifically, for example:
+* you might include the "superclass" struct as a member of the subclass
+  struct;
+* you might use an enum with different variants for the different possible
+  "subclasses".
+
+Usually the answer is obvious: it's unlikely that your Rust code is structured
+in such a way that inheritance would be a good fit anyway.
+
+> I've only missed inheritance when actually implementing languages which
+> themselves have inheritance - MG.
+
+
+## I need a list of nodes which can refer to one another. How?
+
+You can't easily do self-referential data structures in Rust. The usual workaround is to [use an arena](https://manishearth.github.io/blog/2021/03/15/arenas-in-rust/) and replace references from one node to another with node IDs.
+
+An arena is typically a `Vec` (or similar), and the node IDs are a newtype wrapper around a simple integer index.
+
+Obviously, Rust doesn't check that your node IDs are valid. If you don't have proper references, what stops you from having stale IDs?
+
+Arenas are often purely additive, which means that you can add entries but not delete them ([example](https://github.com/Manishearth/elsa/blob/master/examples/mutable_arena.rs)). If you must have an arena which deletes things, then use generational IDs; see the [generational-arena](https://docs.rs/generational-arena/) crate and this [RustConf keynote](https://www.youtube.com/watch?v=aKLntZcp27M) for more details.
+
+If arenas still sound like a nasty workaround, consider that you might choose an arena anyway for other reasons:
+
+* All of the objects in the arena will be freed at the end of the arena's lifetime, instead of during their manipulation, which can give very low latency for some use-cases. [Bumpalo](https://docs.rs/bumpalo/3.6.1/bumpalo/) formalizes this.
+* The rest of your program might have real Rust references into the arena. You can give the arena a named lifetime (`'arena` for example), making the provenance of those references very clear.
+
+## I'm having a miserable time making my data structure. Should I use unsafe?
+
+Low-level data structures are _hard_ in Rust. Arguably, Rust merely makes plain
+all the lifetime and ownership issues which you already had in other languages, but
+the compiler is brutal about it, and you're going to have a bad day.
```

Comment:
Maybe it's just me, but I find fuzzers and VRP reports brutal, and the compiler friendly. :)

I might say:

Low-level data structures are _hard_ in Rust. Rust makes plain all the lifetime and ownership issues that you had in other non-GC languages, at compile time instead of at run time. Even though it can be frustrating, dealing with these bugs right away is ultimately cheaper than finding the bug the hard way: in production.

The Rust compiler goes to a lot of effort to explain the problem precisely, and points to documentation. I think of it as an assistant, helping you prove your assumptions about the lifetimes of objects.

Reply:
I've reworded to this:

```
Low-level data structures are hard in Rust, especially if they're self-
referential. Rust will make visible all sorts of risks of ownership and
shared mutable state which may not be visible in other languages, and
they're hard to solve in low-level data structure code.
```

because I don't 100% agree with you. I think once you get down to high-performance low-level data structures, the borrow checker sometimes genuinely gets in the way with false-positive concerns (as well as protecting you from real errors). Hopefully the above phrasing is a good compromise!